### ISSUE TITLE
Receiving an error entry in the completion queue causes a segfault

### ISSUE THREAD

**Issue by etremel** [Derecho-Project/derecho#141]
The SST polling loop, as implemented in lf.cpp, continuously polls the completion queue using LibFabrics and inserts entries into a vector in PollingData that is indexed by the sender's rank. The sender's rank is supposed to be found in the completion queue entry that LibFabrics returns, in `entry.op_context->ce_idx`. If the completion queue contains an error entry, the entry's `op_context` may be null, in which case `lf_poll_completion()` will return a special value of 0xFFFFFFFF for the sender's rank and the polling loop will discard it.

There appears to be another error case that is not handled properly, however. Some "error entries" in the completion queue appear to contain a non-null `op_context` that nonetheless will contain a garbage value for `ce_idx`. For example, this error sometimes occurs if Derecho is running too slowly or pauses for a while  (e.g. because I'm running it in GDB):

```
[error] fi_cq_readerr() read the following error entry:
[error]       op_context:ce_idx=3241240944,remote_id=32511
[error]       flags=516
[error]       len=139637037206104
[error]       buf=0x1
[error]       data=139637037206104
[error]       tag=139636835876864
[error]       olen=1
[error]       err=5
[error]       prov_errno=12:transport retry counter exceeded
[error]       err_data=0x81
[error]       err_data_size=0
```

Since this error entry has a non-null `op_context`, it is returned to the polling loop, on line 657 of lf.cpp. However, since ce_idx=3241240944 is not a valid index in PollingData's vector of completion entries, this causes a segmentation fault in poll_utils.cpp, line 25.

It seems like bad behavior for errors in the completion queue to cause an unrelated module to segfault, even if the error is not really recoverable and the node will have to shut down anyway. Can we handle completion queue errors with nonsense `op_context`s in a more sensible fashion?

**Comment by KenBirman** [comment]
If the retry counter is exceeded, the node we should shut down is the target node, not the sender…

**Comment by sagarjha** [comment]
@Derecho-Project/core This seems to be a really big issue. And it's not clear how this came about, but I suspect a problem with the libfabric installation on Fractus (and/or the version we recommend with Derecho.
@ellerre (Lorenzo Rosa) also ran into this issue. A very simple test with the SST is failing where pushing a row does not update the remote copy and pushing with completion returns a bad completion entry. When pushing without completion, interestingly, all the subsequent pushes still work.
We should find a fix to this problem soon.

**Comment by etremel** [comment]
Update: This problem is still happening and is present even when not using LibFabrics. In the new Verbs-based version of Derecho, killing a node with ctrl-C will sometimes cause an unrelated node to crash with a segmentation fault in PollingData::insert_completion_entry because `verbs_poll_completion()` returned a completion entry with a garbage value for the "index."

For example, I was running an 8-node group and killed compute17, which should have been a recoverable failure. Instead, compute20 crashed with a segmentation fault in PollingData::insert_completion_entry because it was trying to use 32725 as the index into completion_entries, a vector of length 4. Further investigation revealed that `verbs_poll_completion()` had returned the value `{first = 32725, second = {first = 1927582688, second = -1}}`, which means that it returned on line 584 of verbs.cpp and `sctxt->ce_idx()` was 32725. The last text printed from `verbs_poll_completion()` before the crash was
> got bad completion with status: 5, vendor syndrome: 249

Is there any progress on preventing these bad completions from showing up in the completion queue? If not, could we at least change `verbs_poll_completion()` and/or `PollingData::insert_completion_entry()` to gracefully handle corrupted completion entries rather than causing a hard crash with a segfault?

**Comment by KenBirman** [comment]
From Weijia's work on completion queue overflows, I think we now could make a guess at why this happens: at least in some situations, an error might not be tied to a specific request, but is more general and covers a whole bound queue pair (or even a set of them, when several share one completion queue).  Probably the NIC uses the next available completion to report the error, but doesn't necessarily bother to put valid data in the fields irrelevant to the completion code.

So maybe we need to twist the logic around and first check to see if this is a completion code that specifically is relevant to a single request, and only visit the index field if so.  For other completion codes we could have a generic handler that doesn't look at the other fields at all...

**Comment by songweijia** [comment]
According to RDMAMojo's [ibv_poll_cq document](https://www.rdmamojo.com/2013/02/15/ibv_poll_cq/), the error status 5 sounds right:
`IBV_WC_WR_FLUSH_ERR (5) - Work Request Flushed Error: A Work Request was in process or outstanding when the QP transitioned into the Error State.`
And it also mentioned that `wc.wr_id` is "the 64 bits value that was associated with the corresponding Work Request". In both libfabric and verbs implementation, we put a pointer to `wr_id`, which is suggested as a "context" pointer by libfabric. There are two possibilities:
- As Ken suggested, the error completion is general and irrelevant to the completion code. I didn't find information of that kind in the document yet (Maybe they have some hiding feature not documented?) Verbs do provide asynchronous events which are not associated with any work requests, as completion queue overrun. But those are OOB events, not entries in the completion queue.
- The value of `wc.wr_id` is garbled - Mellanox's driver or hardware behaves differently with the document -- we should verify that. 
- The value of the pointer in `wc.wr_id` is correct, but the memory resources are deallocated. I checked this path multiple times. The pointer either points to a local object in the caller of `resources::post_remote_write_with_completion()`, or  `without_completion_sender_ctxt` member of a `_resource` object. Method `resources::post_remote_write_with_completion()` is only called by `src/core/p2p_connection_manager.cpp:140`(the failure check loop) or `include/derecho/sst/detail/sst_impl.hpp:181`(SST put with completion). Both code pieces poll the completion immediately in the same scope of the pointer's context, meaning the context is valid. The `_resources::without_completion_sender_ctxt` enables the periodical signals for bursty `post_remote_write()` (without completion). The `resources` objects are holden by  `p2p_connection_manager`, which is, of course, valid at that moment.

So, to summarise, I think the second situation is highly possible. Can Edward verify if that is the case (I can offer work with you on that if you need)? If that is true, the cure is, as Ken suggested, to wrist the logic to handle it gracefully. I can simply drop the failed completion without reporting to the sender. The sender, either the failure check loop or `SST::put_with_completion`, will timeout and detect the failure. This applies to both verbs and libfabric implementations.

**Comment by KenBirman** [comment]
One more possibility:  LibFabrics or Derecho managed to somehow reused a verb object while the NIC still considers it active from the prior use.  So, one object was on the NIC queue twice.  Eventually  the NIC reports the old completion, and we destroy the object.  But now then it reports the second completion.  When we dereference the pointer for the second time, it points to garbage. 

A Bloom filter could be an inexpensive way to catch this bug.
